### PR TITLE
Feature/summary page patterns

### PR DESCRIPTION
**Pull request by AndyOGo**
This implements various blocks for summary pages, namely `prolist`, `steps` and `summary`.

## Important
At the moment our steps are numbered hardcoded inline a `<span class="step__bullet">1</span>` tag, both in `progress-tracker` and `steps` block. I would rather utilize CSS based counters with pseudo elements - see http://caniuse.com/#feat=css-counters.
But the need for `counter-reset: counter-name` makes it a bit unobvious.

So I would suggest a `.step__bullet--counter` modifier like:
```css
.step__bullet--counter::before {
  counter-increment: step;
  content: counter(step);
}
```

Each block which includes step would need to reset that `step` counter before it's use as follows:
```css
.progress-tracker__flow {
  counter-reset: step;
}

.steps__flow {
  counter-reset: step;
}
```

@davidknezic @elessar-ch 
What do you think?

### CODE REVIEW

**Comment by sventschui** [comment]
Isn't it possible to have the counter scoped to the `.step` element and reset it there?

**Comment by AndyOGo** [comment]
Nope, if you reset the counter on every step you would get only 1.
reset (0) -> increment (1)
reset (0) -> increment (1)
reset (0) -> increment (1)
and so forth
see https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Lists_and_Counters/Using_CSS_counters

**Comment by AndyOGo** [comment]
Hmh, you made me think again. we could try to utilize the `:first-child` selector

File: docs/page/components/form-and-input-elements/snippets/steps.jade [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+template: demo.jade
+---
+
+include /components/includes/mixins.jade
+
+.steps
+  ul.steps__flow
+    li.steps__step
+      .steps__indicator
+        span.step__bullet 1
```
Comment (AndyOGo) [inline docs/page/components/form-and-input-elements/snippets/steps.jade]: `.step__bullet`is utilized here, but it is not a blue bullet like in the `progress-tracker`.
Therefore it's styling is overwritten in steps.less

I would rather keep them separated.

File: less/style/blocks/steps.less [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    margin-left: 31px;
+    padding-bottom: 20px;
+  }
+
+  .steps__title {
+    margin: 0px;
+    margin-bottom: -8px;
+  }
+
+  .step__bullet {
```
Comment (AndyOGo) [inline less/style/blocks/steps.less]: This overwrites `.step__bullet` scoped in `.steps__step`.
We should rather avoid adjoining classes between different BEM blocks

File: less/style/blocks/steps.less [hunk 3]
```diff
@@ -0,0 +1,66 @@
+@steps-bullet-diameter: 32px;
+
+.steps {
+  margin: 15px;
+}
+
+.steps__flow {
+  counter-reset: step;
+
+  ul& {
```
Comment (AndyOGo) [inline less/style/blocks/steps.less]: shouldn't that list-reset be done in the correct class `.steps__flow`

File: less/style/blocks/summary.less [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+  margin-bottom: 10px;
+}
+
+.summary__items {
+  padding: 0;
+  margin: 0;
+  margin-bottom: 20px;
+  list-style: none;
+
+  .summary__item {
```
Comment (AndyOGo) [inline less/style/blocks/summary.less]: Why do we need nesting here?

File: less/style/blocks/summary.less [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+  &:before {
+    content: "";
+    position: absolute;
+    top: 140px;
+    left: -33px;
+    height: 530px;
+    border-left: 1px @color-gray solid;
+  }
+
+  h3 {
```
Comment (AndyOGo) [inline less/style/blocks/summary.less]: Shouldn't we avoid to style tags and I think we have class `.summary__heading` for this tag.
Do we really need nesting here?

File: less/style/blocks/summary.less [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+      font-size: 18px;
+      font-weight: 600;
+      background: @color-white;
+      color: @color-gray;
+      border: 2px @color-gray solid;
+    }
+
+  }
+
+  ul:first-of-type {
```
Comment (AndyOGo) [inline less/style/blocks/summary.less]: Again this should style `.summary__items` not the `ul` tag.
And nesting selectors is not needed.

File: less/style/blocks/summary.less [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+
+    margin: 0;
+    margin-bottom: 30px;
+    padding: 0;
+
+    li {list-style-type: none;
```
Comment (AndyOGo) [inline less/style/blocks/summary.less]: should be `.summary__item`

**Comment by AndyOGo** [comment]
@FrediBach was there a specific reason to style HTML tags directly in favour of BEM, maybe related to AEM Forms?

**Comment by sventschui** [comment]
Styling html tags isn't favorable within the style guide. If you need this because of technology restriction please bridge that in your app. Use BEM classes in the style guide and apply them as mixinis in your apps less.

**Comment by AndyOGo** [comment]
@sventschui 
I appreciate your feedback

**Comment by AndyOGo** [comment]
@sventschui 
I fixed all BEM related issues.

The only open issue is `.step__bullet` and it's overwrite inside `.steps` block, which I don't like.
Either the `.step` block implements different bullet modifiers (like: `.step__bullet--blue`, `.step__bullet--white`) or each block (`.steps`, `.progress-tracker`, etc.) handles their stepping visualisation by themselves.
If they handle it themselves it would also solve the scattered CSS-Counter solution.

As far as I can tell, the bullets have been removed by design from the `.progress-tracker` anyway and been replaced by blocks...

**Comment by sventschui** [comment]
Thanks @AndyOGo 

Can you check with design whether the bulleted progress tracker can be dropped? If so we should simply remove it :) If not, I need to look into this a bit deeper. Haven't had any time so far

**Comment by AndyOGo** [comment]
@sventschui 
I checked it right now with @anja225. Atm the bulleted progress tracker is in use in several forms, hence we have to keep it for backwards compatibility.
In that case I will refactor the `.steps` block to not utilizing the `.step__bullet` block and instead incorporating it's own implementation.
By doing this we win on both sides, gaining compatibility and keeping concerns separated :)

File: less/style/blocks/prolist.less [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    stroke: @color-gray;
+    width: 0.6em;
+    height: 0.6em;
+  }
+}
+
+.prolist__item__content {
+  margin-top: -1px;
+  margin-left: 40px;
+  font-size: 19px;
```
Comment (AndyOGo) [inline less/style/blocks/prolist.less]: `font-size` is hardcoded several times. shouldn't we utilize mixins for this.
especially concerning mobile?

File: less/style/blocks/steps.less [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+  &::before {
+    position: absolute;
+    counter-increment: steps;
+    content: counter(steps);
+    width: @steps-bullet-diameter;
+    height: @steps-bullet-diameter;
+    border-radius: @steps-bullet-diameter;
+    left: 0;
+    line-height: @steps-bullet-diameter - 2px;
+    font-size: 18px;
```
Comment (AndyOGo) [inline less/style/blocks/steps.less]: again hardcoded `font-size`, `font-weight`

File: less/style/blocks/summary.less [hunk 10]
```diff
@@ -0,0 +1,134 @@
+@import '../variables';
+
+.summary {
+  margin: 15px;
+  font-size: 19px;
+}
+
+.summary__heading {
+  font-size: 19px;
```
Comment (AndyOGo) [inline less/style/blocks/summary.less]: again `font-size` hardcoded

File: less/style/blocks/summary.less [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+      content: '';
+      position: absolute;
+      top: -3px;
+      left: -10px;
+      width: 30px;
+      height: @steps-bullet-diameter;
+      background: #fff;
+    }
+
+    &:after {
```
Comment (AndyOGo) [inline less/style/blocks/summary.less]: duplicate CSS properties like `color`, `background`

File: less/style/blocks/summary.less [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+    list-style-type: none;
+    margin: 5px 0 15px 20px;
+    padding: 0;
+
+    &:before {
+      content: '+';
+      display: inline-block;
+      float: left;
+
+      width: 26px;
```
Comment (AndyOGo) [inline less/style/blocks/summary.less]: could be replaced by a `@variable`

**Comment by AndyOGo** [comment]
@anja225 
I just realised that we have hardcoded font-sizes and font-weight values for the summary component.
Shouldn't those value be defined outside and being responsive?

**Comment by AndyOGo** [comment]
@sventschui @anja225 
Mhm, I see only pixel values for dimension in this whole branch. I thought we want to style everything rem based? Or are the any exceptions?

File: less/style/blocks/prolist.less [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+  font-weight: 200;
+  background: @color-white;
+  color: @color-gray;
+  border: 2px @color-gray solid;
+  text-align: center;
+
+  .icon {
+    fill: @color-gray;
+    stroke: @color-gray;
+    width: 0.6em;
```
Comment (AndyOGo) [inline less/style/blocks/prolist.less]: Why are we using relative `em` units here?

File: less/style/blocks/prolist.less [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+  font-size: 19px;
+}
+
+.prolist--conlist .prolist__item__icon {
+  line-height: 20px;
+
+  .icon {
+    fill: @color-gray;
+    stroke: @color-gray;
+    width: 0.4em;
```
Comment (AndyOGo) [inline less/style/blocks/prolist.less]: Again relative `em` unit

File: less/style/blocks/prolist.less [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+  border-radius: 26px;
+  line-height: 20px;
+  font-size: 24px;
+  font-weight: 200;
+  background: @color-white;
+  color: @color-gray;
+  border: 2px @color-gray solid;
+  text-align: center;
+
+  .icon {
```
Comment (AndyOGo) [inline less/style/blocks/prolist.less]: `.icon` styling is overwritten here...

File: less/style/blocks/prolist.less [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+.prolist__item__content {
+  margin-top: -1px;
+  margin-left: 40px;
+  font-size: 19px;
+}
+
+.prolist--conlist .prolist__item__icon {
+  line-height: 20px;
+
+  .icon {
```
Comment (AndyOGo) [inline less/style/blocks/prolist.less]: And again `.icon` overwrite, not very BEM-like...

File: less/style/blocks/prolist.less [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+  }
+}
+
+.prolist__item__content {
+  margin-top: -1px;
+  margin-left: 40px;
+  font-size: 19px;
+}
+
+.prolist--conlist .prolist__item__icon {
```
Comment (AndyOGo) [inline less/style/blocks/prolist.less]: do we need a `.prolist--conlist` modifier. I would create a separate `.conlist` block for this purpose.

Comment (anja225) [inline less/style/blocks/prolist.less]: We do not need a conlist. It does not make sense.

**Comment by AndyOGo** [comment]
the `.prolist` block is missing a header h3 tag before the list.
Headline
Desktop 24px, demi
Mobile 22px, demi

Text
Mobile 17px, book
Desktop 16px, book

**Comment by anja225** [comment]
Please use a normal "-" as list icon for the step block.

File: docs/page/components/form-and-input-elements/snippets/prolist.jade [hunk 18]
```diff
@@ -0,0 +1,28 @@
...
+---
+
+include /components/includes/mixins.jade
+
+h3.prolist__heading Ihre Vorteile
+
+ul.prolist
+  li.prolist__item
+    .prolist__item__icon
+      +svg('add', ['icons-list__icon', 'icon'])
```
Comment (AndyOGo) [inline docs/page/components/form-and-input-elements/snippets/prolist.jade]: I can't find a `.icons-list__icon` ...
As I understand it this `. prolist__item__icon ` should be utilized at the `svg` tag itself

**Comment by AndyOGo** [comment]
We went through the above issues and fixed them.
@sventschui This should be ready to merge now, but I would be happy if you can double check :)

File: docs/page/components/form-and-input-elements/snippets/steps-realhtml.html [hunk 19]
```diff
@@ -0,0 +1,57 @@
+<div class="row summary--magic">
+
+  <div id="guideContainer-rootPanel-pnl1-axatextdraw___guide-item" class="col-md-12 active">
```
Comment (sventschui) [inline docs/page/components/form-and-input-elements/snippets/steps-realhtml.html]: why the heck id's?

File: docs/page/components/form-and-input-elements/snippets/steps-realhtml.html [hunk 20]
```diff
@@ -0,0 +1,57 @@
...
+
+    <div class="cq-placeholder-guideContainer_47rootPanel_47items_47pnl1_47items_47axatextdraw" id="cq-gen293" style="width: 1400px; height: 0px; font-weight: normal; font-style: normal; font-variant: normal; text-transform: none; text-decoration: none; text-indent: 0px; letter-spacing: normal; word-spacing: normal; line-height: normal; text-align: left; vertical-align: baseline; direction: ltr; opacity: 1; top: auto; right: auto; bottom: auto; left: auto; margin: 0px; padding: 0px; border-style: none; border-width: 0px; position: static; visibility: visible; overflow: visible; white-space: normal; float: none; clear: none; cursor: auto; list-style: disc outside none; background-image: none; background-attachment: scroll; background-color: transparent; background-position: 0% 0%; background-repeat: repeat;"></div>
```
Comment (sventschui) [inline docs/page/components/form-and-input-elements/snippets/steps-realhtml.html]: cq classes should not be persisted in the style guide

File: jquery/summary.js [hunk 21]
```diff
@@ -0,0 +1,61 @@
...
+    this.defaults = {}
+
+    this.options = $.extend({}, this.defaults, options)
+
+    this.init()
+  }
+
+  init() {
+
+    const firstY = this.$element.find('h3').first().offset().top - 14;
```
Comment (sventschui) [inline jquery/summary.js]: finding html tags is a bit weird. Maybe add a `js-...` class? @davidknezic opinion?

File: less/style/blocks/steps.less [hunk 22]
```diff
@@ -0,0 +1,93 @@
...
+.steps__title {
+  margin: 0 0 -8px;
+  .make-font-heading-tertiary;
+}
+
+.steps__description {
+  .make-font-large;
+
+  p + ul {
+    // @todo: this needs to be refactored after we have introduced a normalizing/reseting CSS file
```
Comment (sventschui) [inline less/style/blocks/steps.less]: There is a normalizing less file?!

File: less/style/blocks/summary.less [hunk 23]
```diff
@@ -0,0 +1,131 @@
...
+  margin: 0;
+  margin-bottom: 2px;
+}
+
+.summary__edit {
+  font-size: 16px;
+  font-weight: 600;
+}
+
+.summary--magic {
```
Comment (sventschui) [inline less/style/blocks/summary.less]: isn't there a better name than `magic`? `magic` has a negative touch in source code ;)

File: less/style/blocks/summary.less [hunk 24]
```diff
@@ -0,0 +1,131 @@
...
+  &:before {
+    content: "";
+    position: absolute;
+    top: 140px;
+    left: -33px;
+    height: 530px;
+    border-left: 1px @color-gray solid;
+  }
+
+  .summary__heading {
```
Comment (sventschui) [inline less/style/blocks/summary.less]: Shouldn't nest but rather introduce `.summary__heading--magic`

**Comment by sventschui** [comment]
Made some remarks. Otherwise lgtm

**Comment by AndyOGo** [comment]
@sventschui 
Thanks for your great feedback.
Oh man I guess this .summary--magic stuff is hacky crap for our forms.
I will remove it from the style guide and copy it to our forms hacks less file...

File: less/style/blocks/steps.less (hunk 22 above)
Comment (AndyOGo) [inline less/style/blocks/steps.less]: yes, but it's not imported and does not reset tags like h1-h6, p, etc.

File: jquery/summary.js (hunk 21 above)
Comment (AndyOGo) [inline jquery/summary.js]: I will kill this file too, it is doing whatever crazy thing to fix whatever crazy forms issue utilizing the `.summary--magic` modifier...
This is the wrong place for that hack

**Comment by AndyOGo** [comment]
@sventschui @davidknezic @elessar-ch 
I introduced the new `.richtext` block within this feature. It serves for BEM-less styling of richtext content which we get from WYSIWYG-Editors and where we have no control of HTML `class` attributes.

**Comment by AndyOGo** [comment]
@sventschui @davidknezic @elessar-ch 
Decision made, this will be merged.
Any outstanding issue will be tackled by separate issues.

**Pull request merged**